### ISSUE TITLE
Missing space between operators in MathML rendering mode

### ISSUE THREAD

**Issue by physikerwelt** [mathjax/MathJax#1305]
See http://codepen.io/physikerwelt/pen/wKZQby
further information
https://phabricator.wikimedia.org/T119331
https://de.wikipedia.org/wiki/Portal_Diskussion:Mathematik#Neues_Mathematikrendering

**Comment by pkra** [comment]
Thanks for a the report and live sample. I'm relatively certain this is expected behavior (@dpvc will correct me but probably not before next week). 

The MathML generated by MathJax is, I think, appropriate for the TeX source and Gecko is rendering the MathML correctly. The difference in terms of rendering is due to the fact that MathJax (by default) uses TeX layout rules while Gecko follows MathML layout rules. (I think in this case, it comes down to "add a space after multi-letter mi's" but again @dpvc will know best). 

You can see that MathJax would render the MathML source just like Gecko if you used MathML as an input and configured MathJax to use MathML spacing, see http://docs.mathjax.org/en/latest/options/MathML.html.

I guess one could debate whether the `<mi>dim</mi>` is "correct" for `\dim`. For example, LaTeXML generates `mo` instead which changes the rendering under MathML layout rules (and hence Gecko).

(What throws me off is that WebKit (2.8.5) does add space both after `dim`, `k` and `e`, but that's probably just its incomplete implementation.)

**Comment by pkra** [comment]
I just noticed the precise title of the issue. So again, the problem is that `\dim`  is not seen as an operator (just as a function name). And of course `\mathrm` won't be considered as an operator either.

I've tagges this issue expected behavior since it's about the MathML rendering; feel free to open an issue regarding the TeX to MathML conversion.

**Comment by physikerwelt** [comment]
@pkra With regard to MathmL I was expecting to get <m([io])>ker</m$1> rather than <mi>k<mi><mi>e</mi>.

However, we need to find a pragmatic solution for non professional users that a space is displayed between dim and ker but not between k e r. Would changing the http://docs.mathjax.org/en/latest/options/MathML.html solve the problem, without a high chance of breaking thing somewhere else?

**Comment by pkra** [comment]
>  @pkra With regard to MathmL I was expecting to get `<m([io])>ker</m[io]>` rather than `<mi>k<mi><mi>e</mi>`.

I would not expect that from `\mathrm{ker}` and AFAIK it's not how TeX interprets `\mathrm{ker}` for layout. But, e.g., `\operatorname` would give you `<mi>ker</mi>` (yet still no spacing in MathML layout).

> Would changing the http://docs.mathjax.org/en/latest/options/MathML.html solve the problem, without a high chance of breaking thing somewhere else?

The layout would definitely change in the SVG output, albeit subtly. 

I can't answer if the Wikipedia community would consider this "breaking" but I would guess most users expect LaTeX layout rules, not MathML layout rules, given Wikipedia's use of LaTeX to generate PNGs in the past.

**Comment by dpvc** [comment]
> (I think in this case, it comes down to "add a space after multi-letter mi's" but again @dpvc will know best)

In TeX, the spacing is based on adjacent pairs of elements (each element gets a "TeX class" like BIN for binary operator, OPEN for an open delimiter, REL for a relation, OP for an operator like summation, and so on).  There is a matrix on page 170 of the TeXbook that says what the spacing should be between the various pairs of items.  So an OP followed by another OP gets a small amount of spacing, for example.

In MathML spacing is inserted automatically only for `<mo>` elements (and explicitly by `<mspace>` elements).  The spacing for `<mo>` is based on where the `<mo>` is within its container and on the content of the `<mo>`, as specified in the [operator dictionary](http://www.w3.org/TR/MathML3/appendixc.html).  So two `<mi>` elements in a row should not get any extra spacing in MathML (but may get spacing in TeX, as in the case of two adjacent OP elements as described above).

> Would changing the `useMMLspacing` solve the problem, without a high chance of breaking thing somewhere else?

No, this would cause problems in other areas.  For example, this would change how unary operators are displayed.  Also, operators like `:=` will be spaced incorrectly, since they are rendered by TeX as two separate REL elements, and the saving between REL elements is 0 (but in MathML, the spacing will be the normal relation spacing around each of them).

I think the better solution would be to use invisible apply between the operators, which Firefox will render as a small amount of space, similar to the TeX output.  The TeX input jax does put invisible times following things like `\sin` and `\dim`, but only in certain circumstances, and (unfortunately) that doesn't currently include when `\dim` is followed by another operator.  The heuristic for adding invisible times could be altered to include more situations, and probably should.  I would not change the `<mi>` to `<mo>` as that will cause other spacing problems in native MathML rendering.

Finally, note that TeX already has `\ker` defined as an operator, so one should use `\dim\ker` rather than `\dim\mathrm{ker}`.  Note that `\mathrm{ker}x` has the wrong spacing between the "ker" and "x" in both HTML-CSS output and MathML in Firefox, because the semantics are incorrect.  Note that `\operatorname{ker}` would work correctly, but since `\ker` is already defined, why not just use that?

**Comment by pkra** [comment]
Thanks, @dpvc! 

> The heuristic for adding invisible times could be altered to include more situations, and probably should.

I'll open a new issue for that.

> TeX already has \ker defined as an operator

D'oh! :smiley:

**Comment by physikerwelt** [comment]
@dpvc Thank you very much for your comment. I'm happy if it would work correctly for ```\operatorname{something}```. \ker might be not be whitelisted by texvcjs... but that could be changed.
